### PR TITLE
[SPARK-15776][SQL] Divide Expression inside Aggregation function is casted to wrong type

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

This PR fixes the problem that Divide Expression inside Aggregation function is casted to wrong type, which cause `select 1/2` and `select sum(1/2)`returning different result.

**Before the change:**

```
scala> sql("select 1/2 as a").show()
+---+
|  a|
+---+
|0.5|
+---+

scala> sql("select sum(1/2) as a").show()
+---+
|  a|
+---+
|0  |
+---+

scala> sql("select sum(1 / 2) as a").schema
res4: org.apache.spark.sql.types.StructType = StructType(StructField(a,LongType,true))
```

**After the change:**

```
scala> sql("select 1/2 as a").show()
+---+
|  a|
+---+
|0.5|
+---+

scala> sql("select sum(1/2) as a").show()
+---+
|  a|
+---+
|0.5|
+---+

scala> sql("select sum(1/2) as a").schema
res4: org.apache.spark.sql.types.StructType = StructType(StructField(a,DoubleType,true))
```
## How was this patch tested?

Unit test.

This PR is based on https://github.com/apache/spark/pull/13524 by @Sephiroth-Lin

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2847,4 +2847,15 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
   test("SPARK-15887: hive-site.xml should be loaded") {
     assert(spark.sessionState.newHadoopConf().get("hive.in.test") == "true")
   }
+
+  test("SPARK-15776 Divide expression inside an Aggregation function should not " +
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I think we need some low level unit test instead of end-to-end test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 2]
```diff
@@ -213,7 +213,7 @@ case class Multiply(left: Expression, right: Expression)
 case class Divide(left: Expression, right: Expression)
     extends BinaryArithmetic with NullIntolerant {
 
-  override def inputType: AbstractDataType = NumericType
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: we should also cleanup the `divide` expression to remove code for integral division.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -525,7 +525,7 @@ object TypeCoercion {
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveExpressions {
       // Skip nodes who has not been resolved yet,
       // as this is an extra rule which should be applied at last.
-      case e if !e.resolved => e
+      case e if !e.childrenResolved => e
 
       // Decimal and Double remain the same
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: We can simplify this:
```
case e if !e.childrenResolved => e
case d: Divide if d.dataType.isInstanceOf[IntegralType] => ...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 2 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: Shouldn't we add a dedicated expression for integral division? We currently support `DIV` in our parser: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L954

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 4]
```diff
@@ -213,7 +213,7 @@ case class Multiply(left: Expression, right: Expression)
...
+  override def inputType: AbstractDataType = TypeCollection(DoubleType, DecimalType)
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: Shall we use `FractionalType`, which also includes `FloatType` here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 2 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 4 above)
Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: Float is casted to Double.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: +1

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 1 above)
Comment (clockfly) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: @cloud-fan 

It involves Analyzer, low level unit test is hard to cover that?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: or we can remove this rule entirely, the `ImplicitTypeCasts` can already handle this

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 2 above)
Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: @hvanhovell  Thanks for the catch. I will try to test hive behavior more, like how `select 5 div 3.0` behaves in Hive.

**Comment by yhuai** [comment]
I think we need to explain the inconsistent behavior in the PR description, which is the main reason of making this change. (right now, if we run `select 1/2`, the result is 0.5. However, `select sum(1/2)` will return 0).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 3 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Seems `case d: Divide if d.dataType.isInstanceOf[IntegralType]` is not equivalent with the code that we have at here?

(also , let's avoid of unnecessary changes)

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 1 above)
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: TypeCoercionSuite

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala (hunk 2 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala]: Looks like div is from mysql (see http://dev.mysql.com/doc/refman/5.7/en/arithmetic-functions.html).

```
DIV	Integer division
/	Division operator
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 5]
```diff
@@ -630,6 +640,25 @@ class TypeCoercionSuite extends PlanTest {
...
+
+  test("SPARK-15776 Divide expression's dataType should be casted to Double or Decimal") {
+    val analyzer = new RuleExecutor[LogicalPlan] {
+      override val batches =
+        Seq(Batch("Resolution", FixedPoint(10), FunctionArgumentConversion, Division))
+    }
+
+    // Cast integer to double
+    ruleTest(analyzer, sum(Divide(4, 3)), sum(Divide(Cast(4, DoubleType), Cast(3, DoubleType))))
+    // left expression is already Double, skip
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: what if left expression is int and right expression is double?

**Comment by rxin** [comment]
cc @hvanhovell

**Comment by Sephiroth-Lin** [comment]
LGTM thank you

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala [hunk 6]
```diff
@@ -128,12 +133,27 @@ class ArithmeticExpressionSuite extends SparkFunSuite with ExpressionEvalHelper
...
     }
 
-    DataTypeTestUtils.numericTypeWithoutDecimal.foreach { tpe =>
+    Seq(DoubleType, DecimalType.SYSTEM_DEFAULT).foreach { tpe =>
       checkConsistencyBetweenInterpretedAndCodegen(Divide, tpe, tpe)
     }
   }
 
-  test("/ (Divide) for integral type") {
+  test("SPARK-15776: Divide report unresolved when children's data type is " +
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ArithmeticExpressionSuite.scala]: This suite is to test the correctness, not resolution.
BTW, the resolution logic is defined by `ExpectedInputTypes`, I think we don't need to test it again for divide.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 7]
```diff
@@ -345,4 +345,42 @@ class AnalysisSuite extends AnalysisTest {
 
     assertAnalysisSuccess(query)
   }
+
+  test("SPARK-15776: test whether Divide expression's data type can be deduced correctly by " +
+    "analyzer") {
+    val testRelation = LocalRelation(AttributeReference("a", IntegerType)())
+    assertSameSchema(
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: How about defining a tester function, and use that for the rest of the testing? Something like:
```scala
def assertDivideReturnType(left: Expression, right: Expression, dt: DataType): Unit = {
  assertSameSchema(
   Project(Seq(Alias(sum(Divide(left, right)), "a")()), OneRowRelation),
   StructType(Seq(StructField("a", dt, true))))
}
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 8]
```diff
@@ -199,9 +201,18 @@ class TypeCoercionSuite extends PlanTest {
   }
 
   private def ruleTest(rule: Rule[LogicalPlan], initial: Expression, transformed: Expression) {
+    ruleTest(Seq(rule), initial, transformed)
+  }
+
+  private def ruleTest(
+      rules: Seq[Rule[LogicalPlan]], initial: Expression, transformed: Expression): Unit = {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: NIT: one argument per line

**Comment by clockfly** [comment]
@hvanhovell @cloud-fan  Thanks! Updated.

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by hvanhovell** [comment]
LGTM, pending jenkins

**Comment by cloud-fan** [comment]
merging to master/2.0!

**Comment by rxin** [comment]
@clockfly can you add "Closes # #13524 " to the pr description before we merge.
